TELCODOCS-1879 - Adding docs for PTP Operator T-GM leap seconds feature (#79493)
openshift/openshift-docs: modules/ptp-configuring-dynamic-leap-seconds-handling-for-tgm.adoc

// Module included in the following assemblies:
//
// * networking/ptp/configuring-ptp.adoc

:_mod-docs-content-type: PROCEDURE
[id="ptp-configuring-dynamic-leap-seconds-handling-for-tgm_{context}"]
= Configuring dynamic leap seconds handling for PTP grandmaster clocks

> The PTP Operator container image includes the latest `leap-seconds.list` file  that is available at the time of release.

Author: Thanks @michaelryanpeter I made some updates to expand PTP in various places. The operator name is "PTP Operator" so I am not going to expand the abbreviation in the Operator name.

> The PTP Operator container image includes the latest `leap-seconds.list` file  that is available at the time of release.

Reviewer: SGTM. Thanks for taking the team. Though the changes you made are subtle, I do find it a easier to read as someone unfamiliar with the topic.

// Module included in the following assemblies:
//
// *

:_mod-docs-content-type: PROCEDURE
[id="ptp-configuring-dynamic-leap-seconds-handling-for-tgm_{context}"]
= Configuring dynamic leap seconds handling for PTP grandmaster clocks

The PTP Operator container image includes the latest `leap-seconds.list` file  that is available at the time of release.
You can configure the PTP Operator to automatically update the leap second file by using GPS announcements.

> Leap second information is stored in an automatically generated `leap-configmap` custom resource (CR) in the `openshift-ptp` namespace.
> The PTP Operator mounts the `leap-configmap` CR as a volume in the `linuxptp-daemon` pod that is accessible by the `ts2phc` process.
> 
> If the GPS satellite broadcasts new leap second data, the PTP Operator updates the `leap-configmap` CR with the new data.

Reviewer 2: Technically ConfigMap is not a custom resource (CR). Custom resources are only those defined by a CustomResourceDefinition (CRD). But ConfigMap is a **core** resource.

```suggestion
Leap second information is stored in an automatically generated ConfigMap resource named `leap-configmap` in the `openshift-ptp` namespace.
The PTP Operator mounts the `leap-configmap` as a volume in the `linuxptp-daemon` pod that is accessible by the `ts2phc` process.

If the GPS satellite broadcasts new leap second data, the PTP Operator updates the `leap-configmap` with the new data.
```

The `ts2phc` process picks up the changes automatically.

[NOTE]
====
The following procedure is provided as reference.
The {product-version} version of the PTP Operator enables automatic leap second management by default.
====

.Prerequisites

* You have installed the OpenShift CLI (`oc`).

* You have logged in as a user with `cluster-admin` privileges.

* You have install the PTP Operator and configured a PTP grandmaster clock (T-GM) in the cluster.

.Procedure

. Configure automatic leap second handling in the `phc2sysOpts` section of the `PtpConfig` CR.
Set the following options:
+
[source,yaml]
----
phc2sysOpts: -r -u 0 -m -w -N 8 -R 16 -S 2 -s ens2f0 -n 24 <1>
----
<1> Set `-w` to force `phc2sys` to wait until `ptp4l` has synchronized the system hardware clock before starting its own synchronization process.
+
[NOTE]
====
Previously, you applied an offset adjustment in the `phc2sys` configuration (`-O -37`) to account for historical leap seconds.
This is no longer required.
====

. Configure the Intel e810 NIC to

> enable reading of `NAV-TIMELS` messages from the GPS receiver in the `spec.profile.plugins.e810.ublxCmds` section of the `PtpConfig` CR.

Reviewer 2: The software is not reading the information, it is configuring the GPS module to report it
```suggestion
enable reporting of `NAV-TIMELS` messages by the GPS receiver in the `spec.profile.plugins.e810.ublxCmds` section of the `PtpConfig` CR.
```

For example:
+
[source,yaml]

> ----
> - args: #ubxtool -P 29.20 -p CFG-MSG,1,38,300
>     - "-P"
>     - "29.20"
>     - "-p"
>     - "CFG-MSG,1,38,300"
> ----

Reviewer 2: 300 is an out-of-bounds number, there was a bug that we fixed...
```suggestion
----
- args: #ubxtool -P 29.20 -p CFG-MSG,1,38,248
    - "-P"
    - "29.20"
    - "-p"
    - "CFG-MSG,1,38,248"
----
```

.Verification

> You can validate that the leap-second `ConfigMap` CR has been successfully generated by the PTP Operator.

Reviewer 2: ```suggestion
You can validate that the leap-configmap `ConfigMap` has been successfully generated by the PTP Operator.
```

> Following a leap event announcement sent by GPS, you can check that the PTP Operator changed the local copy of the leap seconds file.

Reviewer 2: It's simple to inspect the configmap, but to inspect the local copy of the leap seconds file the user will need to rsh into the pod, check where the leap seconds file is read from by ts2phc and  inspect it. I just anticipate the question "how do you want us to check the leap seconds file"...
```suggestion
Following a leap event announcement sent by GPS, you can check that the PTP Operator changed the local copy of the leap seconds information by inspecting the leap-configmap.
```
